File: hilla_ride_ios/HelloTukTuk/Core/Services/CustomerLocationPublisher.swift
import CoreLocation
import FirebaseFirestore
import Foundation

/// Publishes the customer's GPS to `users/{uid}` while they have an active ride
/// so the driver can see them live on the map.
@MainActor
final class CustomerLocationPublisher: NSObject, CLLocationManagerDelegate {
    static let shared = CustomerLocationPublisher()

    private let manager = CLLocationManager()
    private let firestore = Firestore.firestore()
    private var activeUserId: String?
    private var lastWriteAt: Date?
    private var lastWritten: CLLocation?

    override init() {
        super.init()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyBest
        manager.distanceFilter = MapPresenceConfig.locationPublishMinMoveMeters
        manager.allowsBackgroundLocationUpdates = false
    }

    func start(for userId: String) {
        activeUserId = userId
        manager.requestWhenInUseAuthorization()
        manager.startUpdatingLocation()
    }

    func stop() {
        activeUserId = nil
        manager.stopUpdatingLocation()
        lastWriteAt = nil
        lastWritten = nil
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        Task { @MainActor in
            await publish(location: location)
        }
    }

    private func publish(location: CLLocation) async {
        guard let userId = activeUserId else { return }
        if let lastWriteAt,
           Date().timeIntervalSince(lastWriteAt) < MapPresenceConfig.locationPublishMinInterval,
           let lastWritten,
           location.distance(from: lastWritten) < MapPresenceConfig.locationPublishMinMoveMeters {
            return
        }

        lastWriteAt = Date()
        lastWritten = location
        try? await firestore.collection("users").document(userId).setData([
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "locationUpdatedAt": FieldValue.serverTimestamp()
        ], merge: true)
    }
}

File: lib/core/services/customer_location_publisher.dart
import 'dart:async';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:geolocator/geolocator.dart';
import 'package:hilla_ride/core/constants/map_presence_config.dart';

/// Publishes the customer's GPS to `users/{uid}` during an active ride so the
/// driver can see them live on the map.
class CustomerLocationPublisher {
  CustomerLocationPublisher({FirebaseFirestore? firestore})
      : _firestore = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _firestore;
  StreamSubscription<Position>? _sub;
  String? _userId;
  DateTime? _lastWriteAt;
  Position? _lastWritten;

  Future<void> start(String userId) async {
    if (_userId == userId && _sub != null) return;
    await stop();
    _userId = userId;

    final permission = await Geolocator.checkPermission();
    if (permission == LocationPermission.denied) {
      await Geolocator.requestPermission();
    }
    final enabled = await Geolocator.isLocationServiceEnabled();
    if (!enabled) return;

    _sub = Geolocator.getPositionStream(
      locationSettings: LocationSettings(
        accuracy: LocationAccuracy.high,
        distanceFilter: MapPresenceConfig.locationPublishMinMoveMeters.round(),
      ),
    ).listen(_publish);
  }

  Future<void> stop() async {
    await _sub?.cancel();
    _sub = null;
    _userId = null;
    _lastWriteAt = null;
    _lastWritten = null;
  }

  Future<void> _publish(Position position) async {
    final userId = _userId;
    if (userId == null) return;

    final now = DateTime.now();
    final lastWriteAt = _lastWriteAt;
    final lastWritten = _lastWritten;
    if (lastWriteAt != null &&
        now.difference(lastWriteAt) <
            MapPresenceConfig.locationPublishMinInterval &&
        lastWritten != null) {
      final moved = Geolocator.distanceBetween(
        lastWritten.latitude,
        lastWritten.longitude,
        position.latitude,
        position.longitude,
      );
      if (moved < MapPresenceConfig.locationPublishMinMoveMeters) {
        return;
      }
    }

    _lastWriteAt = now;
    _lastWritten = position;
    try {
      await _firestore.collection('users').doc(userId).set({
        'latitude': position.latitude,
        'longitude': position.longitude,
        'locationUpdatedAt': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));
    } catch (_) {}
  }
}
